modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/js/categorization/categories/EditCategoriesModal.es.js

```diff
@@ -0,0 +1,365 @@
+				categoryId => initialCategories.indexOf(categoryId) == -1
+			);
+		}
+
+		const removedCategories = initialCategories.filter(
+			category => finalCategories.indexOf(category) == -1
+		);
+
+		fetchCategories(URL_UPDATE_CATEGORIES, append ? 'PATCH' : 'PUT', {
+			documentBulkSelection: {
+				documentIds: fileEntries,
+				selectionScope: {
+					folderId,
+					repositoryId,
+					selectAll
+				}
+			},
+			taxonomyCategoryIdsToAdd: addedCategories,
+			taxonomyCategoryIdsToRemove: removedCategories
+		}).then(() => {
+			onModalClose();
+
+			if (bulkStatusComponent) {
+				bulkStatusComponent.startWatch();
+			}
+		});
+	};
+
+	const isMounted = useIsMounted();
+
+	const parseCategories = categories => {
+		return categories.map(item => {
+			return {
+				label: item.taxonomyCategoryName,
+				value: item.taxonomyCategoryId
+			};
+		});
+	};
+
+	const parseVocabularies = useCallback(vocabularies => {
+		let initialCategories = [];
+		const requiredVocabularies = [];
+		const vocabulariesList = [];
+
+		vocabularies.forEach(vocabulary => {
+			const categories = parseCategories(
+				vocabulary.taxonomyCategories || []
+			);
+
+			const categoryIds = categories.map(item => item.value);
+
+			const obj = {
+				id: vocabulary.taxonomyVocabularyId.toString(),
+				required: vocabulary.required,
+				selectedCategoryIds: categoryIds.join(','),
+				selectedItems: categories,
+				singleSelect: !vocabulary.multiValued,
+				title: vocabulary.name
+			};
+
+			vocabulariesList.push(obj);
+
+			if (vocabulary.required) {
+				requiredVocabularies.push(obj);
+			}
+
+			initialCategories = initialCategories.concat(categoryIds);
+		});
+
+		setInitialCategories(initialCategories);
+		setRequiredVocabularies(requiredVocabularies);
+
+		return vocabulariesList;
+	}, []);
+
+	useEffect(() => {
+		const selection = {
+			documentIds: fileEntries,
+			selectionScope: {
+				folderId,
+				repositoryId,
+				selectAll
+			}
+		};
+
+		const urlCategories = `/bulk/v1.0/sites/${
+			groupIds[0]
+		}/taxonomy-vocabularies/common`;
+
+		Promise.all([
+			fetchCategories(urlCategories, 'POST', selection),
+			fetchCategories(URL_SELECTION, 'POST', selection)
+		]).then(([responseCategories, responseSelection]) => {
+			if (responseCategories && responseSelection) {
+				if (isMounted()) {
+					setLoading(false);
+					setDescription(getDescription(responseSelection.size));
+					setMultiple(fileEntries.length > 1 || selectAll);
+					setVocabularies(
+						parseVocabularies(responseCategories.items || [])
+					);
+				}
+			}
+		});
+	}, [
+		fetchCategories,
+		fileEntries,
+		fileEntries.length,
+		folderId,
+		groupIds,
+		isMounted,
+		parseVocabularies,
+		repositoryId,
+		selectAll
+	]);
+
+	useEffect(() => {
+		const isInvalid = requiredVocabularies.length
+			? requiredVocabularies.some(
+					item => item.required && !item.selectedItems.length
+			  )
+			: false;
+		setIsValid(!isInvalid);
+	}, [requiredVocabularies]);
```

Comment:
Why do we need `useEffect` here? This will run *after rendering* any time that `requiredVocabularies` changes,  and if we change the `isInvalid` state we'll end up doing a double render. Why not just calculate `isInvalid` every time and avoid the double render? If you're worried that computing `isInvalid` is expensive (it seems unlikely to be), you could memoize it with `useMemo` (probably not necessary).

Reply:
Initially I didn't have this in a `useEffect` hook, but when talking with @andresfulla he actually suggested I use it here (although I'm not sure he had the full context - so I'm not going to blame him for helping me out).